### ISSUE TITLE
Compilation tries to install something and fails if not root

### ISSUE THREAD

**Issue by emengd** [owlbarn/eigen#5]
Hi,

Compiling the latest release (0.0.5), both directly and through opam, fails on my system with the following error:

```
+ /home/username/.opam2/base/bin/ocamlfind ocamlmklib -o lib/eigen_stubs -L/usr/local/lib -leigen -lstdc++ lib/eigen_utils_stubs.o lib/ffi_eigen_generated_stub.o
/usr/bin/ld: cannot find -leigen
collect2: error: ld returned 1 exit status
Command exited with code 2.
```

I determined this proceeds from an earlier error:
```
cp: cannot create regular file '/usr/local/lib/libeigen.a': Permission denied
```

Essentially, if I get it right, oasis is trying to install a library system-wide so it can use it for the rest of the build (but it took me a lot of time to understand that this was the problem). This will either fail for lack of permission or silently install something the user didn't ask for, both of which are pretty bad I think.

(I eventually managed to go around it by replacing occurrences of `${libdir}` in `_oasis` with some random directory in my `$HOME`. Unless I missed something, the obvious permanent fix is to use a subdirectory instead, but I don't understand oasis well enough to do that; when I tried it I got an error about hygiene problems.)

**Comment by mseri** [comment]
This will need fixing as it is failing reliably with the new opam2 sandboxing

**Comment by ghennequin** [comment]
I just moved to opam 2.0 today, and ran into the same bug... @mseri any update on that?

**Comment by mseri** [comment]
I am in the middle of a long relocation, should be back on it in a week or two. Can you see if the version currently on master works? (you will need to have libeingen3-dev, or the equivalent for your system, installed)

**Comment by mseri** [comment]
The error you see is due to the opam2 sandboxing. Packages are no longer allowed to write outside certain namespaced folders

**Comment by ghennequin** [comment]
so, I can `git pull` and `make oasis & make & make install` OK

but I'd rather use `opam install eigen`; so temporarily, I manually moved `_build/include/libeigen.a` to `~/.opam/4.06.1+flambda/lib/libeigen.a` and run `opam install eigen` again -- it worked this time.

**Comment by mseri** [comment]
You can use opam pin: `git pull` the latest master and `opam pin add eigen path_to_git_repo` to install using that as the base (you can always go back to the old configuration with `opam pin remove eigen`)

**Comment by ghennequin** [comment]
awesome, thanks

**Comment by ryanrhymes** [comment]
I tried to migrate to jbuilder but failed :(  or maybe I just didn't figure out the right way of compiling C++ code in jbuilder, so help is appreciated. Anyway, it is really bad to first compile a lib from C++ code the copy it into a system folder then compile with ocaml code ...

Currently, only operations of sparse matrices depend on Eigen lib. In the long run, I would like Owl to get rid of eigen lib, not only because it is more difficult to interface to C++, but also it is not so fast really. I think SuiteSparse might be a good candidate to replace Eigen.

**Comment by mseri** [comment]
I'll probably revert to an hybrid between the new approach and the old one, still keeping the artifacts local. I am getting huge troubles due to incompatibilities between the versions of eigen shipped by the various distributions.

**Comment by ryanrhymes** [comment]
I agree, I also noticed this issue. It is better to include one distribution to make sure it compiles. Can you please PR this fix? Many thanks.

**Comment by ryanrhymes** [comment]
All right, I've reverted last commit but still using your fix to copy the .a file to include. Please have a look, in case I break anything, let me know, thanks.

**Comment by mseri** [comment]
I was working on reverting it and integrating it with dune. I’ll rebase on your change!

**Comment by ryanrhymes** [comment]
Awesome, I will leave the rest to you then 👍

**Comment by mseri** [comment]
Probably we should move this discussion to a different issue, however the state of what I have got is the following:

- `libeigen.{a,so}` is built invoking `make` and `make dynamic` via `dune`
- the c stubs are generated with `dune`
- the library is compiled via `dune`

The approach is simpler than doing what [`re2`](https://github.com/janestreet/re2/blob/v0.11/src/re2_c/jbuild) or [`ocmal-yaml`](https://github.com/avsm/ocaml-yaml/blob/master/vendor/dune) are doing, but also more error prone.

The static library seems to work fine, but the dynamic one is not linked properly. I am trying to figure it out still.

As a side notice, the commit removing `-Linclude` did break the build for me

**Comment by Freyr666** [comment]
So I can't install owl due to this issue. Is there any way to bypass eigen installation, or install eigen bindings with opam 2?

**Comment by mseri** [comment]
@ryanrhymes have a look at https://github.com/owlbarn/eigen/pull/7, without it it does not seem to work for me. Then I think we need to actually make a new release of eigen, as the current version cannot work with `opam2` at all and you are forced to pin to the master branch of this repo

**Comment by nilsbecker** [comment]
+1 for fixing eigen in some way. i also just failed to install owl via opam due to this.

**Comment by mseri** [comment]
Version 0.0.6 should avoid this problem and work nicely with opam2. PR pending https://github.com/ocaml/opam-repository/pull/12706

**Comment by mseri** [comment]
Fixed since v0.0.6
